pkg/internal/apis/acme/types_issuer.go

```diff
@@ -213,6 +218,22 @@ type ACMEChallengeSolverHTTP01Ingress struct {
 	IngressTemplate *ACMEChallengeSolverHTTP01IngressTemplate
 }
 
+type ACMEChallengeSolverHTTP01Gateway struct {
+	// Optional service type for Kubernetes solver service
+	// +optional
+	ServiceType corev1.ServiceType `json:"serviceType,omitempty"`
+
+	// The labels to set on HTTPRoute resources to solve ACME
+	// challenges that use this challenge solver.
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
```

Comment:
@jakexks could you maybe explain what this labels field is used for? Is this some kind of selector used by Gateway API?

Reply:
Yep it's a selector for a Gateway in my example in the PR text I set up a Gateway like so:

```
---
apiVersion: networking.x-k8s.io/v1alpha1
kind: GatewayClass
metadata:
  name: haproxy-acmesolver
spec:
  controller: haproxy-ingress.github.io/controller
---
apiVersion: networking.x-k8s.io/v1alpha1
kind: Gateway
metadata:
  name: acmesolver
  namespace: haproxy-ingress
spec:
  gatewayClassName: haproxy-acmesolver
  listeners:
  - protocol: HTTP
    port: 80
    routes:
      kind: HTTPRoute
      selector:
        matchLabels:
          usemeto: solve-challenges
      namespaces:
        from: All
```

This gateway will watch all namespaces for HTTPRoutes with a matching label selector.
e.g. creating a solver like this:

```
    solvers:
    - http01:
        gateway:
          labels:
            usemeto: solve-challenges
```

will create a route acted upon by that Gateway.

Reply:
It might be somewhat confusing because now the `labels` field exists under the `gateway` field.
While the labels are actually the labels of the route.
Maybe you could introduce a `routeTemplate` field (or `HTTPRouteTemplate` field maybe)?
And that could contain info like labels and annotations?

Reply:
identical to the podTemplate and serviceTemplate fields

Reply:
```go
// The labels to set on HTTPRoute resources to solve ACME
// challenges that use this challenge solver.
```

Could we mention that cert-manager will create an HTTPRoute using these labels and that these labels should match the Gateway that is meant to be used for solving the HTTP-01 challenge?

Reply:
Podtemplate was removed for bloating the CRDs, I think a Kyverno mutate policy is a better solution in that case.

Added Mael's docstring as it is much better than mine.